snuba/datasets/dataset.py

```diff
@@ -46,164 +39,36 @@ class Dataset(object):
     manipulate the lower layer objects.
     """
 
+    # TODO: Still not sure if the Dataset class really needs the actual Entity object
+    # or just the name.
     def __init__(
-        self,
-        *,
-        storages: Sequence[Storage],
-        query_plan_builder: ClickhouseQueryPlanBuilder,
-        abstract_column_set: ColumnSet,
-        writable_storage: Optional[WritableStorage],
+        self, *, entities: Sequence[Entity], default_entity: Optional[Entity] = None
     ) -> None:
-        self.__storages = storages
-        self.__query_plan_builder = query_plan_builder
-        self.__abstract_column_set = abstract_column_set
-        self.__writable_storage = writable_storage
+        self.__entities: Mapping[str, Entity] = {
+            get_entity_name(entity): entity for entity in entities
+        }
 
-    def get_extensions(self) -> Mapping[str, QueryExtension]:
-        """
-        Returns the extensions for this dataset.
-        Every extension comes as an instance of QueryExtension.
-        The schema tells Snuba how to parse the query.
-        The processor actually does query processing for this extension.
-        """
-        raise NotImplementedError("dataset does not support queries")
-
-    def get_query_processors(self) -> Sequence[QueryProcessor]:
-        """
-        Returns a series of transformation functions (in the form of QueryProcessor objects)
-        that are applied to queries after parsing and before running them on the storage.
-        These are applied in sequence in the same order as they are defined and are supposed
-        to be stateless.
-        """
-        return []
-
-    def get_abstract_columnset(self) -> ColumnSet:
-        """
-        Returns the abstract query schema for this dataset. This is where Entities
-        will come into play since this method will return the structure of the
-        data model.
-        Now the data model is flat so this is just a simple ColumnSet object. With entities
-        this will be a more complex data structure that defines the schema for each entity
-        and their relations.
-        """
-        # TODO: Make this available to the dataset query processors.
-        return self.__abstract_column_set
+        # TODO: This is a convenience while we slowly migrate everything to Entities. This way
+        # every dataset can have a default entity which acts as a passthrough until we can
+        # migrate the datasets to proper entities.
+        self.__default_entity: Entity = None
+        if default_entity:
+            self.__default_entity = default_entity
+        elif len(entities) == 1:
+            self.__default_entity = entities[0]
 
-    def get_query_plan_builder(self) -> ClickhouseQueryPlanBuilder:
```

Comment:
I think there still needs to be a query plan builder at dataset level. 
Today the query plan builder takes in a logical query (from the client for a dataset) and spits a QueryPlan, which includes a Clickhouse query, the physical query processor to execute and the query execution strategy to run the query.
This is when this haoppens: https://github.com/getsentry/snuba/blob/master/snuba/web/query.py#L109

Now his process is divided into two steps:
part 1
1) we have a logical query that references multiple entities
2) for each one of them we need to select which query processor to run
3) we need to optimize the join to run
----
part 2
4) we need to select the storage (basically what the query plan builder was doing before) for each entity
5) we need to select the query processor to run for each selected storage 
6) we need a strategy that runs the whole query.

So I think there still should be some sort of query plan builder at dataset level that performs part 1 and orchestrates the following steps.

Reply:
Yeah OK this makes sense. I wasn't thinking about joins when I made this change.